// express/lib/controllers/appController.js
import fs from 'fs';
import path from 'path';
import Razorpay from 'razorpay';
import App from '../models/appModel.js';
import nodemailer from 'nodemailer';

// === Razorpay Setup ===
const razorpay = new Razorpay({
  key_id: process.env.RAZORPAY_KEY_ID,
  key_secret: process.env.RAZORPAY_KEY_SECRET
});

// === Submit new app ===
export const submitApp = async (req, res) => {
  try {
    const { appName, websiteUrl, email } = req.body;

    if (!appName || !websiteUrl || !email) {
      return res.status(400).json({ error: 'All fields are required' });
    }

    // Save app entry
    const newApp = new App({
      appName,
      websiteUrl,
      email,
      status: 'pending'
    });

    await newApp.save();

    res.json({
      success: true,
      message: 'App submitted successfully. Please complete payment to generate your APK.',
      appId: newApp._id
    });
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
};

// === Search apps ===
export const searchApps = async (req, res) => {
  try {
    const query = req.query.q || '';
    const apps = await App.find({
      appName: { $regex: query, $options: 'i' }
    });
    res.json(apps);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
};

// === Get single app by ID ===
export const getAppById = async (req, res) => {
  try {
    const app = await App.findById(req.params.id);
    if (!app) return res.status(404).json({ error: 'App not found' });
    res.json(app);
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
};

// === Handle Payment Success + Generate APK ===
export const handlePaymentSuccess = async (req, res) => {
  try {
    const { appId, paymentId, email } = req.body;

    const app = await App.findById(appId);
    if (!app) return res.status(404).json({ error: 'App not found' });

    // Mark payment as complete
    app.status = 'paid';
    app.paymentId = paymentId;

    // === Auto-generate a dummy .apk file ===
    const apkDir = path.join(process.cwd(), 'public', 'uploads');
    if (!fs.existsSync(apkDir)) fs.mkdirSync(apkDir, { recursive: true });

    const apkPath = path.join(apkDir, `${app.appName.replace(/\s+/g, '_')}.apk`);
    fs.writeFileSync(apkPath, 'Fake APK binary data');

    app.downloadLink = `/uploads/${path.basename(apkPath)}`;
    await app.save();

    // === Send Email with Download Link ===
    const transporter = nodemailer.createTransport({
      service: 'gmail',
      auth: {
        user: process.env.EMAIL_USER,
        pass: process.env.EMAIL_PASS
      }
    });

    const mailOptions = {
      from: process.env.EMAIL_USER,
      to: email,
      subject: '✅ Your APK is Ready!',
      html: `
        <h2>Your App Build is Ready!</h2>
        <p>Thank you for your payment. You can download your app here:</p>
        <a href="${process.env.BASE_URL}${app.downloadLink}">Download APK</a>
      `
    };

    await transporter.sendMail(mailOptions);

    res.json({
      success: true,
      message: 'Payment successful, APK generated, and download link sent to email!',
      downloadLink: app.downloadLink
    });
  } catch (err) {
    console.error('Payment success error:', err);
    res.status(500).json({ error: 'Something went wrong: ' + err.message });
  }
};
